Architecture/TodoApp/android/core/src/main/java/com/novoda/todoapp/tasks/presenter/TasksPresenter.java

```diff
@@ -48,41 +56,49 @@ public void startPresenting() {
 
     private void subscribeToSourcesFilteredWith(TasksActionListener.Filter filter) {
         subscriptions.add(
-                getTasksObservableFor(filter)
+                getValidatedTasksEventObservable(filter)
+                        .compose(asData(Tasks.class))
                         .subscribe(tasksObserver)
         );
         subscriptions.add(
-                getTasksEventObservableFor(filter)
+                getValidatedTasksEventObservable(filter)
                         .subscribe(tasksEventObserver)
         );
     }
 
-    private Observable<Tasks> getTasksObservableFor(TasksActionListener.Filter filter) {
-        switch (filter) {
-            case ACTIVE:
-                return tasksService.getActiveTasks();
-            case COMPLETED:
-                return tasksService.getCompletedTasks();
-            case ALL:
-                return tasksService.getTasks();
-            default:
-                throw new IllegalArgumentException("Unkown filter type " + filter);
-        }
+    private Observable<Event<Tasks>> getValidatedTasksEventObservable(TasksActionListener.Filter filter) {
+        return getTasksEventObservableFor(filter)
+                .map(new Func1<Event<Tasks>, Event<Tasks>>() {
+                    @Override
+                    public Event<Tasks> call(Event<Tasks> tasksEvent) {
+                        Iterable<SyncedData<Task>> nonDeletedTasks = Iterables.filter(tasksEvent.data().or(Tasks.empty()).all(), shouldDisplayTask());
```

Comment:
It's a lot of code, I know, but it'd be great if you changed it in some other points too. Will mark them out.